### ISSUE TITLE
Usage of pattern on lhs of assignment results in unsafety check for raw pointer deferences being skipped

### ISSUE THREAD

**Issue by Aiden2207** [rust-lang/rust#107288]
I tried this code:

```rust
fn main() {
    let _ = *(0 as *const ());
}
```
[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6b77d30c8376db7b23571564b51e2d5b)

I expected to see this happen: A compiler error, dereferencing pointers is unsafe, and no unsafe block is provided.
Instead, this happened: 
```
   Compiling playground v0.0.1 (/playground)
warning: dereferencing a null pointer
 --> src/main.rs:2:13
  |
2 |     let _ = *(0 as *const ());
  |             ^^^^^^^^^^^^^^^^^ this code causes undefined behavior when executed
  |
  = note: `#[warn(deref_nullptr)]` on by default

warning: `playground` (bin "playground") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.97s
     Running `target/debug/playground`
```
As far as I can tell, this code erroneously compiles for any `Copy` type, so long as the pattern doesn't bind anything to an identifier. Calls to unsafe functions cause a compiler error as normal.
### Meta
Works on the playground when I tested with stable-1.66.1, beta-1.67.0-beta.10, and nightly-1.69.0 (2023-01-23)

**Comment by steffahn** [comment]
Looks like a duplicate of #79735

**Comment by steffahn** [comment]
Ah, there's also #80059. But that discussion does not encounter the "this code causes undefined behavior" warning either.

**Comment by CAD97** [comment]
That warning feels like it's coming from const evaluation (rvalue promotion). I don't know if it actually is, but I would not be surprised if promotion/promotability is involved here somehow.

**Comment by lukas-code** [comment]
This will be fixed by https://github.com/rust-lang/rust/pull/102256.

**Comment by WaffleLapkin** [comment]
This is in fact fixed by https://github.com/rust-lang/rust/pull/102256 on the latest stable:
```
error[E0133]: dereference of raw pointer is unsafe and requires unsafe function or block
 --> src/main.rs:2:13
  |
2 |     let _ = *(0 as *const ());
  |             ^^^^^^^^^^^^^^^^^ dereference of raw pointer
  |
  = note: raw pointers may be null, dangling or unaligned; they can violate aliasing rules and cause data races: all of these are undefined behavior
```
